oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,313 @@
+package oauth2client
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	// The URN for getting verification token offline
+	oobCallbackUrn = "urn:ietf:wg:oauth:2.0:oob"
+	// The URN for token request grant type jwt-bearer
+	jwtBearerUrn = "urn:ietf:params:oauth:grant-type:jwt-bearer"
+)
+
+// Signer interface to support both rsa and ecdsa sign.
+type signerInterface interface {
+	sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) ([]byte, error)
+}
+type rsaSigner struct {
+	pkey *rsa.PrivateKey
+}
+
+func (s rsaSigner) sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) ([]byte, error) {
+	return s.pkey.Sign(rand, msg, opts)
+}
+
+type ecdsaSigner struct {
+	pkey *ecdsa.PrivateKey
+}
+
+func (s ecdsaSigner) sign(rand io.Reader, msg []byte, opts crypto.SignerOpts) ([]byte, error) {
+	return s.pkey.Sign(rand, msg, opts)
+}
+
+// Default function to handle authorize flow. This hanlder prints the URL on
+// console and reads verification from stdin (i.e. let user paste verification
+// code to console).
+func defaultAuthorizeFlowHandler(authorizeUrl string) string {
+	// Print the url on console, let user authorize and paste the token back.
+	fmt.Printf("Go to the following link in your browser:\n\n   %s\n\n",
+		authorizeUrl)
+
+	fmt.Println("Enter verification code: ")
+	code := ""
+	fmt.Scanln(&code)
+	return code
+}
+
+// Run the three-legged oauth authorize flow.
+func authorizeFlow(secret map[string]interface{}, scopes string, handler func(string) string) string {
+	// Marshaw a url to be printed on console. In web based oauth flow, the
+	// browser should redirect the user to this url
+	paramsMap := map[string]string{
+		"access_type":                 "offline",
+		"auth_provider_x509_cert_url": "",
+		"redirect_uri":                oobCallbackUrn,
+		"response_type":               "code",
+		"client_id":                   "",
+		"scope":                       scopes,
+		"project_id":                  "",
+	}
+
+	params := url.Values{}
+	for key := range paramsMap {
```

Comment:
Why don't we have code like this:

params := ...
for key := range params {
  if val, ok := secret[key]; ok {
    params[key] = val
  }
}

Basically, we just reuse params.